### PR TITLE
Challenge 5

### PR DESCRIPTION
**Pull request by rin-st**
What was changed comparing to sre-1
- branch name - challenge 5 instead of 9
- styles
- added time left for guru when channel is challenged
- show time left only for challenged channels (both rube and guru)
- disabled buttons when it's impossible to make that action. Previously error notifications was shown
- removed error in `timeLeft` solidity code. It polluted the console and was thrown very often

Todos:
- [ ] sometimes(rarely) `userSigner` is `undefined` and last symbol or two doesn't count for voucher. Not sure someone will notice it. Difficult to debug
- [x] fix warning in console `recieved unexpected channel data`
- [x] add total eth locked
- [x] do we need an events page? (there wasn't)
- [x] home page
- [x] images to readme
- [ ] remove Checkpoint 3 code block in `streamer.tsx` before merging
- [x] fix tests
- [ ] fix "not updating on change tab" bug

Screenshots: 
Снимок экрана 2023-07-31 в 19 07 15
Снимок экрана 2023-07-31 в 19 07 08

### CODE REVIEW

**Comment by Pabl0cks** [comment]
Awesome job @rin-st, I've been testing the challenge today and it went really smooth, you did an awesome migration. 

>  do we need an events page? (there wasn't)

During the test I didn't feel the need to have the tab for this Challenge, and I think Block Explorer can help to have a vision of what's happening.

I found some issues difficult to reproduce, will try to reproduce them step by step tomorrow, maybe they're caused by the `received unexpected channel data` warning. Here is everything I saw:

- When I refresh the browser or change to another tab and then go back to `Streamer`, I can't see the channel I had open with Ruber user, nor all the channels I was interacting with (on Guru user). 
- Visual improvement => AutoPay checkbox is not very intuitive when it's on or off since both have the same color. 
- If I close a channel and then open a new one with the same user, the Guru watches again the previous Voucher and characters served for that user. If I try to continue writing in the text field, it throws an error: `Error: invalid BigNumber value...`.  
Could we clean the data for that voucher to solve this?
image
- Hard to reproduce => Sometimes when I perform actions like opening a channel or closing a channel, my screen does not refresh correctly and I still see the old state on Rube. Checking Guru, I see the total eth staked refreshed correctly, but I can't see the number of channels and the active channels correctly, showing less or more channels than the real active ones.  
__If I can reproduce the random one, I'll let you know the exact steps.__
- Blockexplorer has a strange behavior, maybe is because the way channels work, or maybe is a bug on the block explorer component and has nothing to do with the Challenge.  
Seems that is adding "hidden" rows to the blockexplorer, making the real transactions to start moving to deep pages, so as a user you see them dissapearing from screen and if you click on the next page you see them again. Video => https://www.loom.com/share/17924f2d13fe4b65a629225efae0012b

I can work on those TODO items if still pending:

> home page
> images to readme

 GJ again 🙌

**Comment by rin-st** [comment]
Thanks for great review @Pabl0cks !

> When I refresh

Refresh doesn't work for this app and as I remember it wasn't before. If we really need it I think we'll add it in new pr

> Visual improvement => AutoPay checkbox is not very intuitive when it's on or off since both have the same color.

Right is on, left is off as always. I can add "On" and "Off" words near toggle.

> If I close a channel and then open a new one with the same user

You're doing it after refreshing rube page, right?

> I can work on those TODO items if still pending:

I'd be very grateful! Have almost no time next 2-3 days

**Comment by Pabl0cks** [comment]
Thx for the fast responses @rin-st 🙌

> Refresh doesn't work for this app and as I remember it wasn't before. If we really need it I think we'll add it in new pr
> could you add details? just switch tabs and it doesn't work?

Ok! Makes sense. In the deployment I've found it seems it kept the channel open even refreshing, but other stuff wasn't working properly, so I'll investigate a bit more.
By changing tabs I meant clicking to Home and then back to Streamer for example => so same behaviour as Refresh.

> Right is on, left is off as always. I can add "On" and "Off" words near toggle.

Yeah I think it probably won't generate doubts like that. I was expecting something like in this vid (tried changing `toggle-primary` to `toggle-secondary`) but then in dark mode may be confusing, so not the best solution.
https://www.loom.com/share/bfd66a79629c43a89094965989619643

Maybe we can just leave it like that or change the text to "AutoPay off"/ "AutoPay on". Let's hear others feedback before changing anything I would say.

> You're doing it after refreshing rube page, right?

Nope, in that case I wasn't refreshing, the page got "stuck" without changing the visual aspect after clicking the button, altought the transaction executed. Today I couldn't find time to try and reproduce, but tomorrow morning I'll have time.

> I'd be very grateful! Have almost no time next 2-3 days

Sure! Tomorrow morning I'll get on this 🙌

**Comment by Pabl0cks** [comment]
I've just pushed some changes:
- Update Index, Header and MetaHeader with Challenge 5 descriptions and assets.
- Add screenshots to the Readme.
- Add missing MetaHeader to Streamer page, and change checkbox style when checked. If we feel it's worse because in dark mode may be confusing, we can just roll it back to the previous primary. 

Need to review the Readme texts and do a bit more testing, hopefully I can finish my review today 🙌

**Comment by rin-st** [comment]
Thank you @Pabl0cks for your help! 🙏 

fixed `unexpected channel data` bug. Please lmk if you find a way to reproduce others!

Regarding autopay, it seems strange how it looks in dark mode. I'll try another implementation tomorrow

**Comment by Pabl0cks** [comment]
> Regarding autopay, it seems strange how it looks in dark mode. I'll try another implementation tomorrow

Sure! Sorry I know my implementation was very poor 🤦‍♂️

--- 

I think I've finished the test, and managed to find the pattern for the "random" bug. Here is what I found in the last round of tests:

- Played with some old deployments of this challenge in Scaffold eth, and there was no issue when you refreshed the browser, or changed to another section of the website. If you had an open channel you would continue having it in screen, or if you had a closed channel, you would still watch the blocking message in the screen so you wouldn't open a new channel. You can check a couple of Sepolia deployments I found:
   - http://sre-state-channels.surge.sh/ => 30 minute cooldown to withdraw challenged channel, careful :)
   - https://terrible-change.surge.sh/
- This refresh behaviour has those consecuences:
   - If you have an open channel you no longer see it, and have to close it in Debug Contracts tab.
   - If you are the Guru, you stop watching the open channels also.
   - When you have a closed channel, the UI doesn't let you open a new one, until you refresh, then you can open a channel again. This is causing other issues like the one I reported in a previous comment =>  `Error: invalid BigNumber value...`.
- The "random" bug that made the screen not refresh but fired the transaction, is caused when you click an action button and change the web browser tab before letting it update. That fires the transaction but when you go back to the tab, the screen is still the old one. Can check this video: https://www.loom.com/share/f8a5ad7853534cc7bd2b68372a4e4226

Maybe the origin of the first issue (refresh) and the last one is the same? Or could be related to some internal change of SE-2? Just thinking out loud 😅

**Comment by Pabl0cks** [comment]
I've just pushed a few minor tweaks to README and comments.

Added a reminder TODO in PR description to remove the Checkpoint 3 code block so we don't forget 🙌

**Comment by rin-st** [comment]
Thanks! Wanted to do it today but forgot again 😄 

> Played with some old deployments of this challenge in Scaffold eth, and there was no issue when you refreshed the browser, or changed to another section of the website. If you had an open channel you would continue having it in screen, or if you had a closed channel, you would still watch the blocking message in the screen so you wouldn't open a new channel. You can check a couple of Sepolia deployments I found:

It worked for me for some minutes, but after refresh in 10 mins it showed "Open channel" button again. Withdrew my 0.5 sepolia eth from `debug` tab :).

I fixed refresh, so, for now, state of channels (opened/challenged/closed) is saved and it seems
> This refresh behaviour has those consequences:

works as expected. 

But it doesn't save messages (it wasn't worked before too). It's difficult to fix it for rubes because they receive messages from guru and just refreshing rube's page can't help.

Other changed things:
Autopay (secondary color)

https://github.com/scaffold-eth/se-2-challenges/assets/25638585/33901cde-94f3-45f3-a08c-2fea3a541ce6

"Cash out timed out" message
Снимок экрана 2023-08-04 в 15 53 42

and some code refactors

**Comment by Pabl0cks** [comment]
Thx for the quick fixes @rin-st ! 🙌 I've just managed to try the new version.

> I fixed refresh, so, for now, state of channels (opened/challenged/closed) is saved and it seems works as expected

I think Refresh is good enough right now! If it didn't load messages in scaffold-eth version, I think is an edge case and probably shouldn't spend time trying to change it. At least Rubes after Refresh can still see their opened channel, and as soon as Guru adds more text, it loads on the Rube, so it's cool I think.  
The most important thing is fixed with this change, which is that you can't open a 2nd channel after 1st one is closed 💪.

> Autopay (secondary color)

I see it OK!

---

With the last commits I think something broke in the Guru Streamer tab. When you click on "Cash out latest voucher" before it would show notification popups for completed transactions and processed transaction errors for the ones that had some problems. Now is not showing notification popup for completed transactions, and I think is not processing errors, it's showing all the log in the popup.

**The transaction is firing and it's withdrawing correctly, is just the notification popups.**

Before:
image

Now:
image

**Comment by rin-st** [comment]
Returned previous notification

Regarding render bug when switching tabs. For now I don't know how to fix it and what causes it. Probably it's because browsers reduce resources to inactive tabs but not sure. I propose to create an issue for this bug and try to fix it later.

**Comment by Pabl0cks** [comment]
> Returned previous notification

Great! Will test it tonight 🙌

> Regarding render bug when switching tabs. For now I don't know how to fix it and what causes it. Probably it's because browsers reduce resources to inactive tabs but not sure. I propose to create an issue for this bug and try to fix it later.

Good idea, I'll create the issue, right now I think is not a big deal since Refresh works and would fix the not updating issue in most of use cases. 

Let's also get the feedback from others that are able to test the Challenge, maybe I'm the only anxious one changing tabs instantly 😂

**Review by Pabl0cks — APPROVED** [review]
Now is working great @rin-st! The challenge is looking good for me to do the initial merge, Good Job!! 🙌

During my tests I didn't notice the `useSigner is undefined` that you mention in the first TODO. Let's try to get more people to play with the Challenge and receive their feedback.

File: README.md [hunk 1]
```diff
@@ -37,40 +51,236 @@ yarn chain
...
+
+---
+
+## Checkpoint 3: Exchange the Service
+
+Now that the channel is funded, and all participants have observed the funding via the emitted event, we can begin our off-chain exchange of service. We are now working in `packages/nextjs/pages/streamer.tsx`.
+
+Functions of note:
+
+- `provideService`: The Guru sends wisdom over the wire to the client.
```
Comment (technophile-04) [inline README.md:124]: Functions ~of~ to note?

Also, I didn't find `provideService` in `streamer.tsx` did we mean to say `provideWisdom` (this is present int `streamer.tsx`) ? 

If we have mistaken, maybe we update `provideWisdom` -> `provideService` in `streamer.tsx` ?

File: packages/nextjs/pages/streamer.tsx [hunk 2]
```diff
@@ -0,0 +1,447 @@
...
+      const packed = utils.solidityPack(["uint256"], [updatedBalance]);
+      const hashed = utils.keccak256(packed);
+      const arrayified = utils.arrayify(hashed);
+
+      const senderAddress = utils.verifyMessage(arrayified, data.signature);
+
+      if (senderAddress !== clientAddress) {
+        return;
+      }
+      // TODO: end of remove block
```
Comment (technophile-04) [inline packages/nextjs/pages/streamer.tsx:131]: Just marking it here so that remember before merge🙌

**Comment by technophile-04** [comment]
Working great @rin-st, awesome work !!! and thanks @Pabl0cks for great testing !! I think we can are ready to merge as the initial iteration 🙌 Just added a small comment for Readme


> Regarding render bug when switching tabs. For now I don't know how to fix it and what causes it. Probably it's because browsers reduce resources to inactive tabs but not sure. I propose to create an issue for this bug and try to fix it later.

Yeah I did some digging checkout this issue -> https://github.com/facebook/react/issues/16773 , Although that issue was for `concurrent` mode but check first comment of it -> https://github.com/facebook/react/issues/16773#issuecomment-531135047

One quick solution that comes to my mind is doing `window.refresh()` once the transaction is complete 

But I don't think it's a big deal since happens only when you switch tabs **really** fast, but we can maybe just mention in short in the readme to refresh the page in case UI doesn't update refresh the page? 

Nevertheless, I think we can merge this 🙌 !! Really great work Rinnat and Pablo!

**Review by technophile-04 — APPROVED** [review]
Working great @rin-st, awesome work !!! and thanks @Pabl0cks for great testing !! I think we can are ready to merge as the initial iteration 🙌 Just added a small comment for Readme


> Regarding render bug when switching tabs. For now I don't know how to fix it and what causes it. Probably it's because browsers reduce resources to inactive tabs but not sure. I propose to create an issue for this bug and try to fix it later.

Yeah I did some digging checkout this issue -> https://github.com/facebook/react/issues/16773 , Although that issue was for `concurrent` mode but check first comment of it -> https://github.com/facebook/react/issues/16773#issuecomment-531135047

One quick solution that comes to my mind is doing `window.refresh()` once the transaction is complete 

But I don't think it's a big deal since happens only when you switch tabs **really** fast, but we can maybe just mention in short in the readme to refresh the page in case UI doesn't update refresh the page? 

Nevertheless, I think we can merge this 🙌 !! Really great work Rinnat and Pablo!

**Comment by technophile-04** [comment]
> sometimes(rarely) userSigner is undefined and last symbol or two doesn't count for voucher. Not sure someone will notice it. Difficult to debug

This is I think wagmi issue 😅, Personally I didn't get faced it but its similar as mentioned here -> https://github.com/wagmi-dev/wagmi/issues/2784#issue-1815414886

**Comment by rin-st** [comment]
> If we have mistaken, maybe we update provideWisdom -> provideService in streamer.tsx ?

Thanks, updated

**Comment by technophile-04** [comment]
Just back-merge #47 and #46 at [26fa22b](https://github.com/scaffold-eth/se-2-challenges/pull/38/commits/26fa22b8b1dbdccf8ae4a171af74394bb7b7a21b), make sure you do `yarn install` to avoid any error before testing 🙌

**Comment by Pabl0cks** [comment]
> Just back-merge https://github.com/scaffold-eth/se-2-challenges/pull/47 and https://github.com/scaffold-eth/se-2-challenges/pull/46 at [26fa22b](https://github.com/scaffold-eth/se-2-challenges/pull/38/commits/26fa22b8b1dbdccf8ae4a171af74394bb7b7a21b), make sure you do yarn install to avoid any error before testing 🙌

Did a few tests after the back-merge and is working nicely 🙌

**Comment by ZakGriffith** [comment]
Mid way through running the challenge and got stuck here: 
What is the call to action on this line in challenge 5:

📝 Edit packages/nextjs/pages/streamer.tsx to enable the UI button for withdrawals.

In the SE1 challenge there was some clear stuff to uncomment but I am not seeing the path forward here.

**Comment by ZakGriffith** [comment]
> Mid way through running the challenge and got stuck here: What is the call to action on this line in challenge 5:
> 
> 📝 Edit packages/nextjs/pages/streamer.tsx to enable the UI button for withdrawals.
> 
> In the SE1 challenge there was some clear stuff to uncomment but I am not seeing the path forward here.

Figured this out.  The button is already setup and ready to go, so there is no action needed by the user.  I will reword this in the ReadMe to closer align with the process.

**Comment by ZakGriffith** [comment]
I am on board for the merge.  Ran it, it's smooth and I can PR the small ReadMe and Contract changes after.

**Comment by ZakGriffith** [comment]
I'll get on it tomorrow.  Right now it's a bunch of poorly written notes in a Google Doc 😂.  Good work on this challenge folks, really clean and good looking!

**Pull request merged**